stryker-mutator/stryker-js
# Implement hit counter in the jest-runner

This is PR for issue [#3174](https://github.com/stryker-mutator/stryker-js/issues/3174)

Instrumenter
✅ Add jest instrumentation for hit limit and counting hits in root/tasks/instrumenter.js

e2e
✅ add new hit-limit e2e test. Test pass for "npm run test:jest" in e2e/test/hit-limit.

Jest Runner
✅  configure hitLimit
✅  Add Unit test for jest-test-runner
❌ determine if timeout occurred using determineHitLimitReached helper
❌  Add integration test for new behavior

There are two issues I couldn't find solutions to.
- How to use testResources correctly:
With root/tasks/instrument-test-resources.js, I made infinite-loop-instrumented testResources. It adds 1 for namespace(ns).hitCount until it is bigger than hitLimit. I thought the acitveMutantId for this code was 24(I will link with the code below). So I put 24 for activeMutantId and resolved infinite-loop-instrumented file for integration test case but with no luck. I debugged and found hitCounter is 0. I could not find clues for connecting the ns.hitCount with jest-test-runner.,

Also as I work with this, I also found integration test examples with different Ids. I wonder where to find these Ids are connected with. For i.g. the code below uses mutant id 11. But in jasmine2-node-instrumented/src/Circle.js there is no 11.
```
    it('should let mutant 11 survive', async () => {
      const exampleProjectRoot = resolveTestResource('jasmine2-node-instrumented');
      process.chdir(resolveTestResource('jasmine2-node-instrumented'));
      const jestTestRunner = createSut();
      const mutantRunOptions = factory.mutantRunOptions({
        sandboxFileName: require.resolve(path.resolve(exampleProjectRoot, 'src', 'Circle.js')),
      });
      mutantRunOptions.activeMutant.id = '11';
      const runResult = await jestTestRunner.mutantRun(mutantRunOptions);
      assertions.expectSurvived(runResult);
    });
```

- How to use namespace and intrumenterContext correctly:
This is a confirmation question. I implemented instrumentContext for jest-test-runner instead of using process.env.[INSTRUMENT_CONSTANTS], because it only allows string whereas hitCounter and hitLimit is number. In addition, I wanted to use __stryker2__ namespace. But not really sure about this. So, I need your opinion.
